File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [n] * (n + 1)
        dp[0] = 0
        for i in range(1, n + 1):
            curr = i
            for j in range(1, i):
                if i - j * j < 0:
                    break
                curr = min(curr, dp[i - j * j] + 1)
            dp[i] = curr
        return dp[n]

File: 0279-perfect-squares/0279-perfect-squares.java
class Solution {
    public int numSquares(int n) {
        int[] dp = new int[n + 1]; // dp[0] = 0 by default
        for (int i = 1; i <= n; i++) {
            int curr = i;
            for (int j = 1; j * j <= i; j++) {
                int cand = dp[i - j * j];
                if (cand < curr) curr = cand;
            }
            dp[i] = curr + 1;
        }
        return dp[n];
    }
}